File: app/routers/incoming_call.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from datetime import datetime, timedelta
from ..database import get_db
from ..models import Customer, Order, IncomingCall
from ..security import get_current_user
from pydantic import BaseModel

router = APIRouter(prefix="/api", tags=["incoming"])


class IncomingCallPayload(BaseModel):
    phone: str


def _build_customer_data(customer, db):
    orders = db.query(Order).filter(
        Order.customer_id == customer.id
    ).order_by(Order.created_at.desc()).limit(5).all()

    orders_data = []
    for order in orders:
        items = [
            {"product_name": item.product.name, "quantity": item.quantity, "unit_price": item.unit_price}
            for item in order.items
        ]
        orders_data.append({
            "id": order.id,
            "status": order.status,
            "created_at": order.created_at.isoformat(),
            "items": items,
            "total": sum(i["quantity"] * i["unit_price"] for i in items),
        })

    return {
        "id": customer.id,
        "phone": customer.phone,
        "name": customer.name,
        "address": customer.address,
        "note": customer.note,
        "orders": orders_data,
    }


@router.post("/incoming-call")
async def incoming_call(payload: IncomingCallPayload, db: Session = Depends(get_db)):
    phone = payload.phone.strip()

    # Store in DB for polling
    call = IncomingCall(phone=phone)
    db.add(call)
    db.commit()
    db.refresh(call)

    customer = db.query(Customer).filter(Customer.phone == phone).first()
    customer_data = _build_customer_data(customer, db) if customer else None

    return {"ok": True, "customer": customer_data, "found": customer is not None}


@router.get("/incoming-call/pending")
def get_pending_calls(db: Session = Depends(get_db), _: str = Depends(get_current_user)):
    """Frontend polls this every 3s to check for new calls."""
    cutoff = datetime.utcnow() - timedelta(minutes=2)
    calls = db.query(IncomingCall).filter(
        IncomingCall.acknowledged == False,
        IncomingCall.created_at >= cutoff,
    ).order_by(IncomingCall.created_at.desc()).all()

    result = []
    for call in calls:
        customer = db.query(Customer).filter(Customer.phone == call.phone).first()
        result.append({
            "id": call.id,
            "phone": call.phone,
            "created_at": call.created_at.isoformat(),
            "customer": _build_customer_data(customer, db) if customer else None,
        })

    return result


@router.post("/incoming-call/{call_id}/ack")
def acknowledge_call(call_id: int, db: Session = Depends(get_db), _: str = Depends(get_current_user)):
    """Mark a call as acknowledged (modal was shown/closed)."""
    call = db.query(IncomingCall).filter(IncomingCall.id == call_id).first()
    if call:
        call.acknowledged = True
        db.commit()
    return {"ok": True}

File: src/routes/incoming_call.ts
import { Hono } from 'hono'
import { eq, desc, and, gte } from 'drizzle-orm'
import { getDb } from '../db'
import { incomingCalls, customers, orders, orderItems, products } from '../schema'
import type { Env } from '../worker'

const app = new Hono<{ Bindings: Env }>()

async function buildCustomerData(db: ReturnType<typeof getDb>, phone: string) {
  const [customer] = await db.select().from(customers).where(eq(customers.phone, phone)).limit(1)
  if (!customer) return null

  const recentOrders = await db.select({
    id: orders.id, status: orders.status, createdAt: orders.createdAt,
    itemId: orderItems.id, productId: orderItems.productId,
    quantity: orderItems.quantity, unitPrice: orderItems.unitPrice,
    productName: products.name,
  }).from(orders)
    .leftJoin(orderItems, eq(orderItems.orderId, orders.id))
    .leftJoin(products, eq(orderItems.productId, products.id))
    .where(eq(orders.customerId, customer.id))
    .orderBy(desc(orders.createdAt))
    .limit(50)

  // Group by order
  const orderMap = new Map<number, any>()
  for (const row of recentOrders) {
    if (!orderMap.has(row.id)) {
      orderMap.set(row.id, { id: row.id, status: row.status, created_at: row.createdAt, items: [], total: 0 })
    }
    if (row.itemId) {
      const item = { product_name: row.productName, quantity: row.quantity!, unit_price: row.unitPrice! }
      orderMap.get(row.id).items.push(item)
      orderMap.get(row.id).total += row.quantity! * row.unitPrice!
    }
  }

  return {
    id: customer.id, phone: customer.phone, name: customer.name,
    address: customer.address, note: customer.note,
    orders: Array.from(orderMap.values()).slice(0, 5),
  }
}

// Public — APK uses this
app.post('/', async (c) => {
  const db = getDb(c.env.DB)
  const { phone } = await c.req.json()
  const trimmed = phone.trim()

  // Store call in DB
  await db.insert(incomingCalls).values({ phone: trimmed })

  // Broadcast via Durable Object WebSocket
  const hub = c.env.CALL_HUB.get(c.env.CALL_HUB.idFromName('main'))
  const customerData = await buildCustomerData(db, trimmed)
  await hub.fetch(new Request('https://hub/broadcast', {
    method: 'POST',
    body: JSON.stringify({ type: 'incoming_call', phone: trimmed, customer: customerData }),
    headers: { 'Content-Type': 'application/json' },
  }))

  return c.json({ ok: true, customer: customerData, found: !!customerData })
})

// Protected — browser uses these
app.get('/pending', async (c) => {
  const db = getDb(c.env.DB)
  // SQLite stores datetime as "YYYY-MM-DD HH:MM:SS" (space, no T/Z)
  const cutoff = new Date(Date.now() - 2 * 60 * 1000).toISOString().replace('T', ' ').replace('Z', '').slice(0, 19)
  const calls = await db.select().from(incomingCalls)
    .where(and(eq(incomingCalls.acknowledged, false), gte(incomingCalls.createdAt, cutoff)))
    .orderBy(desc(incomingCalls.createdAt))

  const result = await Promise.all(calls.map(async (call) => ({
    id: call.id, phone: call.phone, created_at: call.createdAt,
    customer: await buildCustomerData(db, call.phone),
  })))
  return c.json(result)
})

app.post('/:id/ack', async (c) => {
  const db = getDb(c.env.DB)
  await db.update(incomingCalls).set({ acknowledged: true }).where(eq(incomingCalls.id, Number(c.req.param('id'))))
  return c.json({ ok: true })
})

export default app
